### DISCUSSION TITLE
Android clients are still broken

### DISCUSSION THREAD

**Discussion by sairuk** [dani-garcia/vaultwarden#5329] [General]
I'm am currently running the 1.32.7 docker image with multiple (5) android users on client versions 2024.10.0 (11270) or 2024.12.0 (19597). 

All users lose access to their vaults with the official bitwarden android client, behaviour began while we were on 1.32.5 and earlier android client versions and has persisted through 1.32.6 to 1.32.7 vw.

Reviewing previous PRs, issues etc and attempting the fixes like clearing cache/data/app, rebooting, rolling keys etc has no effect to restoring access however a reinstall will allow initial sync of the vault but eventually (could be days) it will stop working again with "We were unable to process your request. Please try again or contact us"

The only particulars I have in place for my main instance is
- restricted access to our internal network (so when one of my users is out and they happen to have lost access to their vault, they have to wait until they return to reinstall and regain access)
- a custom port
- multiple shared vaults

any debugging tips, thoughts on rectifying this?

**Comment by BlackDex** [comment]
Use `adb logcat` to see what the issue is. 
It does sometimes help to fully clear all data of the app, and fully login again.

**Reply by sairuk** [reply]
Only seeing ui events dropped to logcat, filtering app by uid with `adb logcat --uid=$(adb shell pm list package -U com.x8bit.bitwarden | sed 's/.*uid://')`

```
12-29 11:17:46.368 13393 13393 W WindowOnBackDispatcher: Set 'android:enableOnBackInvokedCallback="true"' in the application manifest.
12-29 11:17:46.369 13393 13441 W Parcel  : Expecting binder but got null!
12-29 11:17:46.400 13393 13437 E OpenGLRenderer: Unable to match the desired swap behavior.
12-29 11:21:04.856 13393 13437 E OpenGLRenderer: Unable to match the desired swap behavior.
12-29 11:21:04.867 13393 13393 W WindowOnBackDispatcher: OnBackInvokedCallback is not enabled for the application.
12-29 11:21:04.867 13393 13393 W WindowOnBackDispatcher: Set 'android:enableOnBackInvokedCallback="true"' in the application manifest.
12-29 11:21:04.886 13393 13437 E OpenGLRenderer: Unable to match the desired swap behavior.
```

however as i have mentioned this morning in https://github.com/dani-garcia/vaultwarden/issues/5237#issuecomment-2564563757 i noted the last sync is reporting "blank" when the app is in a failed state whereas previously it was populated after reinstall/setup. 

hitting try again on the error messages or selecting 'sync' from the dot menu results in this log output in vaultwarden logs
```
[2024-12-29 11:33:31.449][request][INFO] GET /api/sync
[2024-12-29 11:33:31.492][response][INFO] (sync) GET /api/sync?<data..> => 200 OK
```

if trying to force a sync through the "other" menu it appears to hang indefinitely although the output on the vaultwarden side still appears to be the same
```
[2024-12-29 11:35:38.953][request][INFO] GET /api/sync
[2024-12-29 11:35:38.995][response][INFO] (sync) GET /api/sync?<data..> => 200 OK
```

**Reply by sairuk** [reply]
also a full clear of the app (cache/data) and re-login/setup does not change the behaviour, only a complete app reinstall does.

**Reply by BlackDex** [reply]
If that is the only error log from the Bitwarden app, then I'm not sure what the issue is. It doesn't indicate anything with the sync is self. 

Also, is the sync works right after a clean install, then it's also probably not something in Vaultwarden, else that first sync would have failed too.

Also check your reverse proxy logs, and check in the `/admin/diagnostics` page of there are any issues reported.

**Reply by sairuk** [reply]
no reverse proxy here atm, direct to the container on a custom port mapped with docker ... which is (was) fine for my current use case. vw trace logging doesn't show anything of concern in particular either

also /admin/diagnostics reports all OK


i've converted to the fdroid repo now to test older app versions
- 2024.10.1 (19388) cannot complete initial sync
- 2024.11.0 (19421) cannot complete initial sync
- 2024.11.7 (19528) cannot complete initial sync

back to trying out play store releases
- 2024.10.0 (11270) completes initial sync
- 2024.12.0 (19597), joined beta and updated, vault is immediately broken
- 2024.12.0 (19597), uninstall/reinstall cannot complete initial sync

must've gotten my wires crossed somewhere as the beta will not sync successfully after reinstallation i thought it was at one point. 2024.10.0 (current play store version) must have been the version i reinstalled with sync initially working.

**Reply by BlackDex** [reply]
But what does show up in the adb logs? 
Are you 100% sure you are running v1.32.7?
Are you using a self signed cert, those can cause some issues. 

What happens with a new clean account?
Can you use chrome to visit the web-vault?

What if you try to download a dev version and check the adb logs? You can find it here https://github.com/bitwarden/android/actions/runs/12380700849 for example, there are newer versions though.

**Reply by sairuk** [reply]
adb logs, i didn't collect for all those version tests, google is now installing 12.0 even though i've left the beta so i'll need to wait for whatever on their end, behaviour is the same so we can assume the output is near enough to what i've already posted, but i'll take a look at a dev build.

And yes I made sure I was on 1.32.7 according to diagnostics / docker tag prior to starting this discussion
![image](https://github.com/user-attachments/assets/538c3423-8c2c-4e14-ac9e-e5b6cce47156)

certificate is a letsencrypt wildcard cert nothing special there, offering up fullchain and privkey and yeh all is fine with web-vault, no issues nor are there issues with the firefox extension.

I tried converting one of the vaults from PBKDF2 to Argon2Id to see if it would trigger some behaviour, however no luck.

i'm going to start fresh and see what happens, going to be a few days work because we have multiple users / shared vaults here and i need to coordinate with folks, migrate them to a new instance, export/import etc if it appears to be working.

I also have another instance here that is broken owned by a single user on their own net so i'll work with them to migrate first, see if a fresh setup gets anywhere.

I spoke to someone here who has 1.32.7 and app version 2024.11.5 working fine.

**Reply by BlackDex** [reply]
Strange, because if the sync fails, it normally shows an exception in de adb logcat output, which might be cryptic, but could help finding the issue. I didn't meant creating a whole new Vaultwarden instance, but a new account without any passwords or organizations, and see if that works, and what happens of you add passwords or invite them to an org etc...

**Reply by sairuk** [reply]
It is fine I knew what you meant and i will give a new user a go as well. Based on recent findings I think that may work out ok, still a bit of work but not terrible.

I've migrated the single user instance to a new instance and now the latest server (1.32.7) / client (2024.12.0) software (so far) is happy enough with each other while prior to this with their old db the android app at the same version failed (this is on Android 12). Perhaps it is some additional legacy db entry breaking stuff for the new android clients.

Also while going through some devices here I have found one of my users is successfully connecting to my existing 1.32.7 instance with 2024.12.0 (10597) on Android 11 / Pixel 2, they were only added quite recently when we were on 1.32.6 while my Android 14 / Pixel 4a on any client after 2024.10.0 is broken. This was the first user added when we migrated from pass which was a while ago now but what versions of vw we were on at the time escapes me atm.

**Reply by BlackDex** [reply]
The android error during the sync will be useful to see if we still need to fix something wrong on the db side. Else a har/sync response would be useful to try and figure out what could be wrong.

**Reply by sairuk** [reply]
I've fixed this now, the sync wasn't failing. So no expected logcat messages. It appears bw was failing to successfully parse the data it received from vw.

The newer bitwarden clients appear to fall over on any invalid match entries, in my case it appears I had uri match entries that were an empty string `""` instead of `null` as evidenced in my exported data. Annoying that it kills the entire vault access without any reasonable error messages. At a guess the problematic entries may have been the result of a previous batch migration from pass a while back but I no longer have access to the old data to confirm.

To repair this I exported the vaults for the affected users, processed them with python to set any match with a uri to `null` then imported the data into a new instance/user and found that the latest android client was happy with the vault. I then purged the users vaults on the older instance and imported their corrected data and it is happy now.

I also had an org breaking 2024.12.0, due to what I expect was also missing match entries that I manually fixed, however this was not enough to get sync working. Turning off access to collections one by one was enough to troubleshoot the issue as once the offending collection access was disabled the client could sync again. To fix this I ended up exporting and importing the data into a new org, however no changes were made to the exported data prior to import, the android client successfully synced the new org/collections.

That is now server 1.32.7 and client 2024.12.0 working for me.

Thanks for you help looking into this issue, if you have any further questions let me know.

**Reply by BlackDex** [reply]
Thanks for reporting back, i have fixed this issue via #5332
